README.md
# bottom_dialog

Bottom dialog is a picker dialog for pick with single or multiply chooses. with support of adding icons or and search.

<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/1.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/2.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/3.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/4.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/5.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/6.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/7.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/8.png?raw=true" width="250" height="500">
<img src="https://github.com/georgesamirmansour/bottom_dialog/blob/master/screenShot/9.png?raw=true" width="250" height="500">


## Getting Started



## Examples: BottomModelExample
```groovy
 List<BottomModel> list = [
      BottomModel(
          name: 'Camera',
          id: 1,
          selected: true,
          imageWidget: Icon(Icons.camera),
          sortId: 1),
      BottomModel(
          name: 'Gallery',
          id: 2,
          selected: false,
          imageWidget: Icon(Icons.image),
          sortId: 1)
    ];
```
## Examples: ShowDialog
```groovy
    ShowBottomDialog(
      context: context,
      bottomModelList: list,
      // show or hide image inside list
      showImage: true,
      // list item text style
      itemTextStyle: TextStyle(color: Colors.black87, fontSize: 18.0),
      // title text in center of header
      titleText: Text(
        'Select which',
        style: TextStyle(color: Colors.black, fontSize: 22.0),
      ),
      // save word text, only showed when multiSelect = true,
      saveWord: 'Save',
      // save text style
      saveTextStyle: TextStyle(color: Colors.blueAccent, fontSize: 18.0),
      // background color
      widgetColor: Colors.white,
      // allow use search inside dialog
      useSearch: true,
      // search text
      searchWord: 'Search...',
      // search text style
      searchTextStyle: TextStyle(color: Colors.blueAccent, fontSize: 16.0),
      // search hint text style
      searchHintTextStyle: TextStyle(color: Colors.grey, fontSize: 16.0),

      // custom widget for search icon as prefix
      prefixSearchIcon: Icon(
        Icons.search,
        color: Colors.blueAccent,
      ),
      // if true => return onSave function, else return onItemSelected
      multiSelect: true,
      onSave: (List<BottomModel> returnedList) {},
      // return
      onItemSelected: (BottomModel item) {},
      // allow to show shadow at top and bottom of dialog
      hasShadow: true,
      // make custom dialog
      customBoxShadow: [
        BoxShadow(
          color: Colors.blueAccent,
          blurRadius: 5.0,
          spreadRadius: 5.0,
          offset: Offset(2.0, 2.0), // shadow direction: bottom right
        )
      ],
      // divider after header (cancel, title text...)
      hasDividerAfterHeader: true,
      // divider color
      dividerColor: Colors.blueGrey,
      // cancel text
      cancelWord: 'Cancel',
      // cancel text style , will throw error if it's null
      cancelTextStyle: TextStyle(color: Colors.red, fontSize: 16.0),
      // dialog top left and right radius
      dialogRadius: 30.0,
    );
```
License
--------
MIT License

Copyright (c) 2021 George Samir

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
## [1.0.0]
* initial release.
## [1.0.1]
* fixed image in readme.
## [1.0.2]
* fixed image in readme.
## [1.0.3]
* refactored model id to string.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example/main.dart';



void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/bottom_dialog.dart
import 'package:bottom_dialog/bottom_model.dart';
import 'package:flutter/material.dart';

class BottomDialog extends StatefulWidget {
  final bool showImage,
      useSearch,
      multiSelect,
      hasShadow,
      hasDividerAfterHeader;
  final List<BottomModel> bottomModelList;
  final Function(List<BottomModel> bottomModelList)? onSave;
  final Function(BottomModel bottomModelItem)? onItemSelected;
  final Text titleText;
  final Color widgetColor;
  final double dialogRadius;
  final List<BoxShadow>? customBoxShadow;
  final Color? dividerColor;
  final TextStyle? searchTextStyle,
      searchHintTextStyle,
      saveTextStyle,
      cancelTextStyle;
  final String? searchWord, saveWord, cancelWord;
  final Widget? prefixSearchIcon;
  final TextStyle itemTextStyle;

  BottomDialog(
      {required this.showImage,
      required this.useSearch,
      required this.multiSelect,
      required this.bottomModelList,
      required this.titleText,
      required this.itemTextStyle,
      this.onSave,
      this.onItemSelected,
      this.widgetColor = Colors.white,
      this.dialogRadius = 30.0,
      this.customBoxShadow,
      this.hasShadow = true,
      this.dividerColor,
      this.searchTextStyle,
      this.searchHintTextStyle,
      this.searchWord,
      this.prefixSearchIcon,
      this.saveTextStyle,
      this.cancelTextStyle,
      this.saveWord,
      this.cancelWord,
      this.hasDividerAfterHeader = false});

  @override
  _BottomDialogState createState() => _BottomDialogState();
}

class _BottomDialogState extends State<BottomDialog> {
  List<BottomModel> _tempList = [];

  @override
  void initState() {
    _tempList = widget.bottomModelList;
    sortedList(_tempList);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return _customScaffold;
  }

  Widget get _customScaffold =>
      Scaffold(backgroundColor: Colors.transparent, body: _customSafeArea);

  Widget get _customSafeArea => SafeArea(
        child: _customColumn,
      );

  Widget get _customColumn => Column(
        children: [_expandedTopSpace, _screenDesign],
      );

  Widget get _expandedTopSpace =>
      Expanded(flex: 1, child: SizedBox(height: 150.0));

  Widget get _screenDesign => Container(
        decoration: _customBoxDecoration,
        padding: _customContainerVerticalPadding,
        child: _customContainerColumn,
      );

  BoxDecoration get _customBoxDecoration => BoxDecoration(
      color: widget.widgetColor,
      borderRadius: _customBorderRadius,
      boxShadow: _customBoxShadow);

  BorderRadius get _customBorderRadius => BorderRadius.only(
      topLeft: Radius.circular(widget.dialogRadius),
      topRight: Radius.circular(widget.dialogRadius));

  List<BoxShadow>? get _customBoxShadow => _whichShadowToReturn();

  List<BoxShadow>? _whichShadowToReturn() {
    if (widget.hasShadow) {
      if (widget.customBoxShadow == null)
        return _defaultBoxShadow;
      else
        return widget.customBoxShadow!;
    } else
      return null;
  }

  List<BoxShadow> get _defaultBoxShadow => [
        BoxShadow(
          color: _grayColor,
          blurRadius: 5.0,
          spreadRadius: 5.0,
          offset: Offset(2.0, 2.0), // shadow direction: bottom right
        )
      ];

  Widget get _customContainerColumn => Column(
        children: [
          _titleRow,
          _customHeightSizedBox,
          if(widget.hasDividerAfterHeader)
          _customDivider,
          _customHeightSizedBox,
          if (widget.useSearch) _customSearchFiled,
          if (widget.useSearch) _customHeightSizedBox,
          _customList,
          _customHeightSizedBox,
        ],
      );

  Widget get _titleRow => Row(
        mainAxisAlignment: widget.multiSelect
            ? MainAxisAlignment.spaceBetween
            : MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          _customHeaderExpanded(_cancelTextButton),
          _customHeaderExpanded(widget.titleText),
          _customHeaderExpanded(
              widget.multiSelect ? _saveTextButton : Container())
        ],
      );

  Expanded _customHeaderExpanded(Widget widget) {
    return Expanded(
      child: widget,
      flex: 1,
    );
  }

  TextButton get _cancelTextButton => TextButton(
      onPressed: () {
        Navigator.pop(context);
      },
      child: _cancelText);

  Text get _cancelText => Text(
        widget.cancelWord!,
        style: widget.cancelTextStyle,
      );

  TextButton get _saveTextButton => TextButton(
      onPressed: () {
        Navigator.pop(context);
        widget.onSave!(widget.bottomModelList);
      },
      child: _saveText);

  Text get _saveText => Text(
        widget.saveWord!,
        style: widget.saveTextStyle,
      );

  Widget get _customSearchFiled => Padding(
    padding: _customContainerHorizontalPadding,
    child: TextField(
          style: widget.searchTextStyle,
          onChanged: (value) {
            _doSearchOnList(value);
          },
          decoration: InputDecoration(
              hintText: widget.searchWord,
              hintStyle: widget.searchHintTextStyle,
              prefixIcon: widget.prefixSearchIcon),
        ),
  );

  void _doSearchOnList(String searchValue) {
    setState(() {
      if (searchValue.isEmpty)
        _tempList = widget.bottomModelList;
      else
        _tempList = widget.bottomModelList
            .where((element) =>
                element.name.toLowerCase().contains(searchValue.toLowerCase()))
            .toList();
      sortedList(_tempList);
    });
  }

  void sortedList(List<BottomModel> list){
   list.sort((a, b) => a.sortId!.compareTo(b.sortId!));
  }

  Widget get _customList => Container(
        height: _tempList.length > 2
            ? MediaQuery.of(context).size.height / 3
            : null,
        padding: _customContainerHorizontalPadding,
        child: ListView.builder(
          itemBuilder: (context, index) => _customItem(_tempList[index]),
          shrinkWrap: true,
          itemCount: _tempList.length,
        ),
      );

  _customItem(BottomModel item) {
    return InkWell(
      onTap: () {
        _onTapOnItem(item);
      },
      child: _listItem(item),
    );
  }

  Widget _listItem(BottomModel item) {
    return Padding(
      padding: _customItemRowPadding,
      child: _itemRow(item),
    );
  }

  Row _itemRow(BottomModel item) {
    return Row(
      children: [
        if (widget.multiSelect) _customCheckBox(item),
        if (widget.multiSelect) _customWidthSizedBox,
        if (widget.showImage) _customImageSizedBox(item.imageWidget),
        if (widget.showImage) _customWidthSizedBox,
        _customItemText(item.name)
      ],
    );
  }

  Checkbox _customCheckBox(BottomModel item) {
    return Checkbox(
        value: item.selected,
        onChanged: (value) {
          item.selected = value!;
          _changeSelectedItem(item);
        });
  }

  Text _customItemText(String text) {
    return Text(
      text,
      style: widget.itemTextStyle,
    );
  }

  SizedBox _customImageSizedBox(Widget? imageWidget) {
    return SizedBox(
      width: 30.0,
      height: 30.0,
      child: imageWidget,
    );
  }

  void _onTapOnItem(BottomModel item) {
    if (!widget.multiSelect)
      doSingleSelect(item);
    else
      _changeSelectedItem(item);
  }

  void doSingleSelect(BottomModel item) {
    Navigator.pop(context);
    widget.onItemSelected!(item);
  }

  void _changeSelectedItem(BottomModel item) {
    widget.bottomModelList
        .singleWhere((element) => item.id == element.id)
        .selected = !item.selected;
    setState(() {
      _tempList.singleWhere((element) => item.id == element.id).selected =
          !item.selected;
    });
  }

  Color get _grayColor => Color.fromARGB(255, 112, 112, 112);

  EdgeInsets get _customContainerVerticalPadding =>
      EdgeInsets.only(top: 15.0, bottom: 15.0);

  EdgeInsets get _customContainerHorizontalPadding =>
      EdgeInsets.only(left: 30.0, right: 30.0);

  SizedBox get _customHeightSizedBox => SizedBox(
        height: 8.0,
      );

  Divider get _customDivider => Divider(
        height: 1.0,
        color: widget.dividerColor == null ? _grayColor : widget.dividerColor,
      );

  SizedBox get _customWidthSizedBox => SizedBox(
        width: 10.0,
      );

  EdgeInsets get _customItemRowPadding =>
      EdgeInsets.only(left: 15.0, bottom: 10.0, right: 15.0, top: 10.0);
}


lib/show_bottom_dialog.dart
import 'package:bottom_dialog/bottom_dialog.dart';
import 'package:flutter/material.dart';
import 'bottom_model.dart';

export 'bottom_model.dart';

class ShowBottomDialog {
  final bool showImage,
      useSearch,
      multiSelect,
      hasShadow,
      hasDividerAfterHeader;
  final List<BottomModel> bottomModelList;
  final Function(List<BottomModel> bottomModelList)? onSave;
  final Function(BottomModel bottomModelItem)? onItemSelected;
  final Text titleText;
  final Color widgetColor;
  final double dialogRadius;
  final List<BoxShadow>? customBoxShadow;
  final Color? dividerColor;
  final TextStyle? searchTextStyle,
      searchHintTextStyle,
      saveTextStyle,
      cancelTextStyle;
  final String? searchWord, saveWord, cancelWord;
  final Widget? prefixSearchIcon;
  final TextStyle itemTextStyle;
  final BuildContext context;

  ShowBottomDialog({
    required this.showImage,
    required this.useSearch,
    required this.multiSelect,
    required this.bottomModelList,
    required this.titleText,
    required this.itemTextStyle,
    required this.context,
    this.onSave,
    this.onItemSelected,
    this.widgetColor = Colors.white,
    this.dialogRadius = 30.0,
    this.customBoxShadow,
    this.hasShadow = true,
    this.dividerColor,
    this.searchTextStyle,
    this.searchHintTextStyle,
    this.searchWord,
    this.prefixSearchIcon,
    this.saveTextStyle,
    this.cancelTextStyle,
    this.saveWord,
    this.cancelWord,
    this.hasDividerAfterHeader = false,
  }) {
    _showBottomDialog(BottomDialog(
      showImage: showImage,
      useSearch: useSearch,
      multiSelect: multiSelect,
      bottomModelList: bottomModelList,
      titleText: titleText,
      itemTextStyle: itemTextStyle,
      dialogRadius: dialogRadius,
      customBoxShadow: customBoxShadow,
      dividerColor: dividerColor,
      hasShadow: hasShadow,
      onItemSelected: onItemSelected,
      onSave: onSave,
      prefixSearchIcon: prefixSearchIcon,
      searchHintTextStyle: searchHintTextStyle,
      searchTextStyle: searchTextStyle,
      searchWord: searchWord,
      widgetColor: widgetColor,
      cancelTextStyle: cancelTextStyle,
      cancelWord: cancelWord,
      saveTextStyle: saveTextStyle,
      saveWord: saveWord,
      hasDividerAfterHeader: hasDividerAfterHeader,
    ));
  }

  void _showBottomDialog(Widget widget) {
    Navigator.of(context).push(PageRouteBuilder(
        opaque: false,
        transitionsBuilder: (context, animation, secondaryAnimation, child) {
          var begin = Offset(0.0, 1.0);
          var end = Offset.zero;
          var tween = Tween(begin: begin, end: end);
          var offsetAnimation = animation.drive(tween);
          return SlideTransition(
            position: offsetAnimation,
            child: child,
          );
        },
        pageBuilder: (context, animation, secondaryAnimation) => widget));
  }
}


example/lib/main.dart
import 'package:bottom_dialog/bottom_model.dart';
import 'package:flutter/material.dart';
import 'package:bottom_dialog/show_bottom_dialog.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Bottom dialog Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Bottom dialog Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'click on floating action button to show dialog',
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _showDialog,
        tooltip: 'Increment',
        child: Text(
          'show dialog',
          textAlign: TextAlign.center,
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  void _showDialog() {
    List<BottomModel> list = [
      BottomModel(
          name: 'One',
          id: '1',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 1),
      BottomModel(
          name: 'Two',
          id: '2',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 2),
      BottomModel(
          name: 'Three',
          id: '3',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 3),
      BottomModel(
          name: 'Four',
          id: '4',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 4),
      BottomModel(
          name: 'Five',
          id: '5',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 5),
      BottomModel(
          name: 'Seven',
          id: '7',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 7),
      BottomModel(
          name: 'Six',
          id: '6',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 6),
      BottomModel(
          name: 'Eight',
          id: '8',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 8),
      BottomModel(
          name: 'Nine',
          id: '9',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 9),
      BottomModel(
          name: 'Ten',
          id: '10',
          selected: false,
          imageWidget: Icon(Icons.confirmation_number_sharp),
          sortId: 10),
    ];
    ShowBottomDialog(
      context: context,
      bottomModelList: list,
      // show or hide image inside list
      showImage: true,
      // list item text style
      itemTextStyle: TextStyle(color: Colors.black87, fontSize: 18.0),
      // title text in center of header
      titleText: Text(
        'Select which',
        style: TextStyle(color: Colors.black, fontSize: 22.0),
      ),
      // save word text, only showed when multiSelect = true,
      saveWord: 'Save',
      // save text style
      saveTextStyle: TextStyle(color: Colors.blueAccent, fontSize: 18.0),
      // background color
      widgetColor: Colors.white,
      // allow use search inside dialog
      useSearch: true,
      // search text
      searchWord: 'Search...',
      // search text style
      searchTextStyle: TextStyle(color: Colors.blueAccent, fontSize: 16.0),
      // search hint text style
      searchHintTextStyle: TextStyle(color: Colors.grey, fontSize: 16.0),

      // custom widget for search icon as prefix
      prefixSearchIcon: Icon(
        Icons.search,
        color: Colors.blueAccent,
      ),
      // if true => return onSave function, else return onItemSelected
      multiSelect: true,
      onSave: (List<BottomModel> returnedList) {},
      // return
      onItemSelected: (BottomModel item) {},
      // allow to show shadow at top and bottom of dialog
      hasShadow: true,
      // make custom dialog
      customBoxShadow: [
        BoxShadow(
          color: Colors.blueAccent,
          blurRadius: 5.0,
          spreadRadius: 5.0,
          offset: Offset(2.0, 2.0), // shadow direction: bottom right
        )
      ],
      // divider after header (cancel, title text...)
      hasDividerAfterHeader: true,
      // divider color
      dividerColor: Colors.blueGrey,
      // cancel text
      cancelWord: 'Cancel',
      // cancel text style , will throw error if it's null
      cancelTextStyle: TextStyle(color: Colors.red, fontSize: 16.0),
      // dialog top left and right radius
      dialogRadius: 30.0,
    );
  }
}
